src/docfx/build/redirection/RedirectionProvider.cs

```diff
@@ -201,14 +206,38 @@ private static void GenerateRedirectionRules(ErrorBuilder errors, string docsetP
 
             foreach (var item in redirections.Concat(renames))
             {
-                if (item.SourcePath.IsDefault || string.IsNullOrEmpty(item.RedirectUrl))
+                if ((item.SourcePath.IsDefault && item.SourcePathFromRoot.IsDefault) || string.IsNullOrEmpty(item.RedirectUrl))
                 {
                     // Give a missing-attribute warning when source_path or redirect_url not specified
-                    errors.Add(Errors.JsonSchema.MissingAttribute(item.RedirectUrl, "source_path or redirect_url"));
+                    errors.Add(Errors.JsonSchema.MissingAttribute(item.RedirectUrl, "(source_path or source_path_from_root) or redirect_url"));
+                    continue;
+                }
+                else if (!item.SourcePath.IsDefault && !item.SourcePathFromRoot.IsDefault)
+                {
+                    // TODO: "source_path" and "source_path_from_root" exist at the same time.
                     continue;
                 }
 
-                var sourcePath = Path.GetRelativePath(docsetPath, Path.Combine(basedir, item.SourcePath));
+                string sourcePath;
+
+                if (!item.SourcePath.IsDefault)
+                {
+                    // TODO: syntax check
+                    sourcePath = Path.GetRelativePath(docsetPath, Path.Combine(basedir, item.SourcePath));
+                }
+                else
+                {
+                    // TODO: syntax check
+                    var sourcePathRelativeToRepoRoot = item.SourcePathFromRoot.Value.Substring(1);
+                    if (repository != null)
+                    {
+                        sourcePath = Path.GetRelativePath(docsetPath, Path.Combine(repository.Path, sourcePathRelativeToRepoRoot));
+                    }
+                    else
```

Comment:
>else [](start = 20, length = 4)

what is the case when repository == null?

Reply:
For spec test.